In [1]:
import pandas as pd
import os
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt


**Sanctions List - data exploration and descriptives**

*Analysis Plan:*


1.   Load data
2.   get overview of the entire file (column names, size of the df, frequencies, dtypes)
3. get number of missingness for the Designation/ Entry into Force Dates
4. Transform Dtypes
5. Filter dataframe for missing Designation entries only to decide on relevance
6. get descriptives & plot data







In [2]:
os.getcwd()

'C:\\Users\\cecilianatalie.strom\\GIGA\\INSA Project - Dokumente\\Sanction_Data\\EU'

In [3]:
df=pd.read_csv('C:\\Users\\cecilianatalie.strom\\GIGA\\INSA Project - Dokumente\\Sanction_Data\\EU\\eu_sanctions_historic.csv',sep=";",low_memory=False)
df

fileGenerationDate  Entity_LogicalId Entity_EU_ReferenceNumber  \
0             29/11/2023                13                  EU.27.28   
1             29/11/2023                13                  EU.27.28   
2             29/11/2023                13                  EU.27.28   
3             29/11/2023                13                  EU.27.28   
4             29/11/2023                13                  EU.27.28   
...                  ...               ...                       ...   
30221         29/11/2023            158039                EU.10793.9   
30222         29/11/2023            158039                EU.10793.9   
30223         29/11/2023            158039                EU.10793.9   
30224         29/11/2023            158039                EU.10793.9   
30225         29/11/2023            158039                EU.10793.9   

      Entity_UnitedNationId Entity_DesignationDate  \
0                       NaN                    NaN   
1                       NaN                    NaN   
2                       NaN                    NaN   
3                       NaN                    NaN   
4                       NaN                    NaN   
...                     ...                    ...   
30221                   NaN             2023-05-26   
30222                   NaN             2023-05-26   
30223                   NaN             2023-05-26   
30224                   NaN             2023-05-26   
30225                   NaN             2023-05-26   

             Entity_DesignationDetails  \
0                                  NaN   
1                                  NaN   
2                                  NaN   
3                                  NaN   
4                                  NaN   
...                                ...   
30221  UN Security Council designation   
30222  UN Security Council designation   
30223  UN Security Council designation   
30224  UN Security Council designation   
30225  UN Security Council designation   

                                           Entity_Remark Entity_SubjectType  \
0                                 (UNSC RESOLUTION 1483)                  P   
1                                 (UNSC RESOLUTION 1483)                  P   
2                                 (UNSC RESOLUTION 1483)                  P   
3                                 (UNSC RESOLUTION 1483)                  P   
4                                 (UNSC RESOLUTION 1483)                  P   
...                                                  ...                ...   
30221  (Date of UN designation: 2023-05-26)\n(Designa...                  P   
30222  (Date of UN designation: 2023-05-26)\n(Designa...                  P   
30223  (Date of UN designation: 2023-05-26)\n(Designa...                  P   
30224  (Date of UN designation: 2023-05-26)\n(Designa...                  P   
30225  (Date of UN designation: 2023-05-26)\n(Designa...                  P   

      Entity_SubjectType_ClassificationCode Entity_Regulation_Type  ...  \
0                                    person             regulation  ...   
1                                    person             regulation  ...   
2                                    person             regulation  ...   
3                                    person             regulation  ...   
4                                    person             regulation  ...   
...                                     ...                    ...  ...   
30221                                person              amendment  ...   
30222                                person              amendment  ...   
30223                                person              amendment  ...   
30224                                person              amendment  ...   
30225                                person              amendment  ...   

      Citizenship_LogicalId Citizenship_RegulationLanguage Citizenship_Remark  \
0                       NaN                            N

In [4]:
df.shape
#shows us the numer of rows & columns - 118 columns, we need to filter that!

(30226, 118)

In [5]:
#get all the missing values
is_na = df.isna().sum()
is_na

fileGenerationDate                               0
Entity_LogicalId                                 0
Entity_EU_ReferenceNumber                        0
Entity_UnitedNationId                        29420
Entity_DesignationDate                        5138
                                             ...  
Citizenship_Regulation_PublicationDate       28367
Citizenship_Regulation_EntryIntoForceDate    28367
Citizenship_Regulation_NumberTitle           28367
Citizenship_Regulation_Programme             28367
Citizenship_Regulation_PublicationUrl        28367
Length: 118, dtype: int64

In [6]:
#get the percentages of missingsness
values_list = list()
cols_list = list() #creating empty lists

for col in df.columns:
    pct_missing= np.mean(df[col].isnull())*100
    cols_list.append(col)
    values_list.append(pct_missing)

pct_missing_df =pd.DataFrame()
pct_missing_df["col"] = cols_list
pct_missing_df["pct_missing"] = values_list
pct_missing_df

col  pct_missing
0                           fileGenerationDate     0.000000
1                             Entity_LogicalId     0.000000
2                    Entity_EU_ReferenceNumber     0.000000
3                        Entity_UnitedNationId    97.333422
4                       Entity_DesignationDate    16.998610
..                                         ...          ...
113     Citizenship_Regulation_PublicationDate    93.849666
114  Citizenship_Regulation_EntryIntoForceDate    93.849666
115         Citizenship_Regulation_NumberTitle    93.849666
116           Citizenship_Regulation_Programme    93.849666
117      Citizenship_Regulation_PublicationUrl    93.849666

[118 rows x 2 columns]

|index|col|pct\_missing|
|---|---|---|
|0|fileGenerationDate|0\.0|
|1|Entity\_LogicalId|0\.0|
|2|Entity\_EU\_ReferenceNumber|0\.0|
|3|Entity\_UnitedNationId|97\.33342155759942|
|4|Entity\_DesignationDate|16\.998610467809172|
|5|Entity\_DesignationDetails|97\.71058029511018|
|6|Entity\_Remark|11\.437173294514656|
|7|Entity\_SubjectType|0\.0|
|8|Entity\_SubjectType\_ClassificationCode|0\.0|
|9|Entity\_Regulation\_Type|0\.0|
|10|Entity\_Regulation\_OrganisationType|0\.0|
|11|Entity\_Regulation\_PublicationDate|0\.0|
|12|Entity\_Regulation\_EntryIntoForceDate|0\.0|
|13|Entity\_Regulation\_NumberTitle|0\.0|
|14|Entity\_Regulation\_Programme|0\.0|
|15|Entity\_Regulation\_PublicationUrl|0\.0|
|16|NameAlias\_LastName|80\.711969827301|
|17|NameAlias\_FirstName|80\.74505392708264|
|18|NameAlias\_MiddleName|86\.97809832594456|
|19|NameAlias\_WholeName|27\.013167471713096|
|20|NameAlias\_NameLanguage|54\.80050287831668|
|21|NameAlias\_Gender|70\.85952491232713|
|22|NameAlias\_Title|97\.85284192417124|
|23|NameAlias\_Function|88\.23860252762522|
|24|NameAlias\_LogicalId|27\.013167471713096|
|25|NameAlias\_RegulationLanguage|27\.013167471713096|
|26|NameAlias\_Remark|94\.98445047310263|
|27|NameAlias\_Regulation\_Type|27\.013167471713096|
|28|NameAlias\_Regulation\_OrganisationType|27\.013167471713096|
|29|NameAlias\_Regulation\_PublicationDate|27\.013167471713096|
|30|NameAlias\_Regulation\_EntryIntoForceDate|27\.013167471713096|
|31|NameAlias\_Regulation\_NumberTitle|27\.013167471713096|
|32|NameAlias\_Regulation\_Programme|27\.013167471713096|
|33|NameAlias\_Regulation\_PublicationUrl|27\.013167471713096|
|34|Address\_City|96\.20525375504533|
|35|Address\_Street|97\.07867398928074|
|36|Address\_PoBox|99\.77171971150665|
|37|Address\_ZipCode|99\.0306358763978|
|38|Address\_Region|99\.52358896314432|
|39|Address\_Place|99\.7948785813538|
|40|Address\_AsAtListingTime|95\.06716072255674|
|41|Address\_ContactInfo|98\.10097267253359|
|42|Address\_CountryIso2Code|95\.06716072255674|
|43|Address\_CountryDescription|95\.06716072255674|
|44|Address\_LogicalId|95\.06716072255674|
|45|Address\_RegulationLanguage|95\.06716072255674|
|46|Address\_Remark|98\.84205650764243|
|47|Address\_Regulation\_Type|95\.06716072255674|
|48|Address\_Regulation\_OrganisationType|95\.06716072255674|
|49|Address\_Regulation\_PublicationDate|95\.06716072255674|
|50|Address\_Regulation\_EntryIntoForceDate|95\.06716072255674|
|51|Address\_Regulation\_NumberTitle|95\.06716072255674|
|52|Address\_Regulation\_Programme|95\.06716072255674|
|53|Address\_Regulation\_PublicationUrl|95\.06716072255674|
|54|BirthDate\_BirthDate|91\.07060146893403|
|55|BirthDate\_Day|91\.07390987891219|
|56|BirthDate\_Month|91\.00112485939258|
|57|BirthDate\_Year|88\.66207900483028|
|58|BirthDate\_YearRangeFrom|99\.83127109111362|
|59|BirthDate\_YearRangeTo|99\.82796268113545|
|60|BirthDate\_Circa|88\.1658175081056|
|61|BirthDate\_CalendarType|88\.1658175081056|
|62|BirthDate\_ZipCode|100\.0|
|63|BirthDate\_Region|98\.91814993714021|
|64|BirthDate\_Place|98\.82882286772977|
|65|BirthDate\_City|94\.24005822801561|
|66|BirthDate\_CountryIso2Code|88\.1658175081056|
|67|BirthDate\_CountryDescription|88\.1658175081056|
|68|BirthDate\_LogicalId|88\.1658175081056|
|69|BirthDate\_RegulationLanguage|88\.1658175081056|
|70|BirthDate\_Remark|97\.9322437636472|
|71|BirthDate\_Regulation\_Type|88\.1658175081056|
|72|BirthDate\_Regulation\_OrganisationType|88\.1658175081056|
|73|BirthDate\_Regulation\_PublicationDate|88\.1658175081056|
|74|BirthDate\_Regulation\_EntryIntoForceDate|88\.1658175081056|
|75|BirthDate\_Regulation\_NumberTitle|88\.1658175081056|
|76|BirthDate\_Regulation\_Programme|88\.1658175081056|
|77|BirthDate\_Regulation\_PublicationUrl|88\.1658175081056|
|78|Identification\_Number|95\.90749685701053|
|79|Identification\_Diplomatic|95\.90749685701053|
|80|Identification\_KnownExpired|95\.90749685701053|
|81|Identification\_KnownFalse|95\.90749685701053|
|82|Identification\_ReportedLost|95\.90749685701053|
|83|Identification\_RevokedByIssuer|95\.90749685701053|
|84|Identification\_IssuedBy|99\.91398134056773|
|85|Identification\_IssuedDate|99\.26884139482564|
|86|Identification\_ValidFrom|99\.89082247072058|
|87|Identification\_ValidTo|99\.55998147290413|
|88|Identification\_LatinNumber|100\.0|
|89|Identification\_NameOnDocument|99\.97353272017469|
|90|Identification\_TypeCode|95\.90749685701053|
|91|Identification\_TypeDescription|95\.90749685701053|
|92|Identification\_Region|99\.85773837093893|
|93|Identification\_CountryIso2Code|95\.90749685701053|
|94|Identification\_CountryDescription|100\.0|
|95|Identification\_LogicalId|95\.90749685701053|
|96|Identification\_RegulationLanguage|95\.90749685701053|
|97|Identification\_Remark|97\.65764573545954|
|98|Identification\_Regulation\_Type|95\.90749685701053|
|99|Identification\_Regulation\_OrganisationType|95\.90749685701053|
|100|Identification\_Regulation\_PublicationDate|95\.90749685701053|
|101|Identification\_Regulation\_EntryIntoForceDate|95\.90749685701053|
|102|Identification\_Regulation\_NumberTitle|95\.90749685701053|
|103|Identification\_Regulation\_Programme|95\.90749685701053|
|104|Identification\_Regulation\_PublicationUrl|95\.90749685701053|
|105|Citizenship\_Region|100\.0|
|106|Citizenship\_CountryIso2Code|93\.8496658505922|
|107|Citizenship\_CountryDescription|100\.0|
|108|Citizenship\_LogicalId|93\.8496658505922|
|109|Citizenship\_RegulationLanguage|93\.8496658505922|
|110|Citizenship\_Remark|99\.91398134056773|
|111|Citizenship\_Regulation\_Type|93\.8496658505922|
|112|Citizenship\_Regulation\_OrganisationType|93\.8496658505922|
|113|Citizenship\_Regulation\_PublicationDate|93\.8496658505922|
|114|Citizenship\_Regulation\_EntryIntoForceDate|93\.8496658505922|
|115|Citizenship\_Regulation\_NumberTitle|93\.8496658505922|
|116|Citizenship\_Regulation\_Programme|93\.8496658505922|
|117|Citizenship\_Regulation\_PublicationUrl|93\.8496658505922|

In [7]:
#get a list of all columns
list1 = df.columns.to_list ()
list1

['fileGenerationDate',
 'Entity_LogicalId',
 'Entity_EU_ReferenceNumber',
 'Entity_UnitedNationId',
 'Entity_DesignationDate',
 'Entity_DesignationDetails',
 'Entity_Remark',
 'Entity_SubjectType',
 'Entity_SubjectType_ClassificationCode',
 'Entity_Regulation_Type',
 'Entity_Regulation_OrganisationType',
 'Entity_Regulation_PublicationDate',
 'Entity_Regulation_EntryIntoForceDate',
 'Entity_Regulation_NumberTitle',
 'Entity_Regulation_Programme',
 'Entity_Regulation_PublicationUrl',
 'NameAlias_LastName',
 'NameAlias_FirstName',
 'NameAlias_MiddleName',
 'NameAlias_WholeName',
 'NameAlias_NameLanguage',
 'NameAlias_Gender',
 'NameAlias_Title',
 'NameAlias_Function',
 'NameAlias_LogicalId',
 'NameAlias_RegulationLanguage',
 'NameAlias_Remark',
 'NameAlias_Regulation_Type',
 'NameAlias_Regulation_OrganisationType',
 'NameAlias_Regulation_PublicationDate',
 'NameAlias_Regulation_EntryIntoForceDate',
 'NameAlias_Regulation_NumberTitle',
 'NameAlias_Regulation_Programme',
 'NameAlias_Regula

In [8]:
#describe the individual columns
print(df["Entity_DesignationDate"].describe())
print(df["Entity_Regulation_PublicationDate"].describe())
print(df["Entity_Regulation_EntryIntoForceDate"].describe())

count          25088
unique           345
top       2022-02-23
freq            1513
Name: Entity_DesignationDate, dtype: object
count          30226
unique           303
top       2022-09-15
freq            2023
Name: Entity_Regulation_PublicationDate, dtype: object
count          30226
unique           308
top       2022-09-16
freq            2023
Name: Entity_Regulation_EntryIntoForceDate, dtype: object


In [9]:
#and get the numbers of missingness for specific variables
print(df["Entity_DesignationDate"].isnull().sum())
print(df["Entity_Regulation_PublicationDate"].isnull().sum())
print(df['Entity_Regulation_EntryIntoForceDate'].isnull().sum())

5138
0
0


we can see, that the Entity Designation Date column has 5138 missing values, lets investiagte them further

In [10]:
#first we have to check the data types and potentially transform them
df.dtypes

fileGenerationDate                           object
Entity_LogicalId                              int64
Entity_EU_ReferenceNumber                    object
Entity_UnitedNationId                        object
Entity_DesignationDate                       object
                                              ...  
Citizenship_Regulation_PublicationDate       object
Citizenship_Regulation_EntryIntoForceDate    object
Citizenship_Regulation_NumberTitle           object
Citizenship_Regulation_Programme             object
Citizenship_Regulation_PublicationUrl        object
Length: 118, dtype: object

In [11]:
#change NaN values to None
df = df.replace(np.nan,None)
#create an additional variable indicating if Entity_DesignationDate is missing or not
def test_func(df):
    """ Test Function for generating new value"""
    if df['Entity_DesignationDate'] == None :
              return 0
    else:
      return 1

df["missing_designation"] = df.apply(test_func, axis=1)
print(df)

      fileGenerationDate  Entity_LogicalId Entity_EU_ReferenceNumber  \
0             29/11/2023                13                  EU.27.28   
1             29/11/2023                13                  EU.27.28   
2             29/11/2023                13                  EU.27.28   
3             29/11/2023                13                  EU.27.28   
4             29/11/2023                13                  EU.27.28   
...                  ...               ...                       ...   
30221         29/11/2023            158039                EU.10793.9   
30222         29/11/2023            158039                EU.10793.9   
30223         29/11/2023            158039                EU.10793.9   
30224         29/11/2023            158039                EU.10793.9   
30225         29/11/2023            158039                EU.10793.9   

      Entity_UnitedNationId Entity_DesignationDate  \
0                      None                   None   
1                      None

In [12]:
df["missing_designation"].describe()

count    30226.000000
mean         0.830014
std          0.375627
min          0.000000
25%          1.000000
50%          1.000000
75%          1.000000
max          1.000000
Name: missing_designation, dtype: float64

In [13]:
#now only show values with missing Designation date and inspect
df_missing = df.query("missing_designation == 0")
df_missing

fileGenerationDate  Entity_LogicalId Entity_EU_ReferenceNumber  \
0             29/11/2023                13                  EU.27.28   
1             29/11/2023                13                  EU.27.28   
2             29/11/2023                13                  EU.27.28   
3             29/11/2023                13                  EU.27.28   
4             29/11/2023                13                  EU.27.28   
...                  ...               ...                       ...   
30195         29/11/2023            157762               EU.10781.42   
30196         29/11/2023            157762               EU.10781.42   
30197         29/11/2023            157762               EU.10781.42   
30198         29/11/2023            157762               EU.10781.42   
30199         29/11/2023            157762               EU.10781.42   

      Entity_UnitedNationId Entity_DesignationDate Entity_DesignationDetails  \
0                      None                   None                      None   
1                      None                   None                      None   
2                      None                   None                      None   
3                      None                   None                      None   
4                      None                   None                      None   
...                     ...                    ...                       ...   
30195                  None                   None                      None   
30196                  None                   None                      None   
30197                  None                   None                      None   
30198                  None                   None                      None   
30199                  None                   None                      None   

                Entity_Remark Entity_SubjectType  \
0      (UNSC RESOLUTION 1483)                  P   
1      (UNSC RESOLUTION 1483)                  P   
2      (UNSC RESOLUTION 1483)                  P   
3      (UNSC RESOLUTION 1483)                  P   
4      (UNSC RESOLUTION 1483)                  P   
...                       ...                ...   
30195                    None                  E   
30196                    None                  E   
30197                    None                  E   
30198                    None                  E   
30199                    None                  E   

      Entity_SubjectType_ClassificationCode Entity_Regulation_Type  ...  \
0                                    person             regulation  ...   
1                                    person             regulation  ...   
2                                    person             regulation  ...   
3                                    person             regulation  ...   
4                                    person             regulation  ...   
...                                     ...                    ...  ...   
30195                            enterprise              amendment  ...   
30196                            enterprise              amendment  ...   
30197                            enterprise              amendment  ...   
30198                            enterprise              amendment  ...   
30199                            enterprise              amendment  ...   

      Citizenship_RegulationLanguage Citizenship_Remark  \
0                               None               None   
1                               None               None   
2                               None               None   
3                               None               None   
4                                 EN               None   
...                              ...                ...   
30195                           None               None   
30196                           None               None   
30197                           None               None   
30198                           None               None   
30199

In [14]:
#save df as separate csv files to inspect manually
df_missing.to_csv('eu_missing_designation.csv')
df.to_csv('eu_sanctions_complete.csv')

**Data Exploration & Visualization**

In [15]:
#plot the number of designations

#distribution of entity types

#frequency of dates/ months/ year

**Create final dataset**

In [16]:
#filter to only retrieve variables needed

#change dtypes

#rename columns
